backstage/community-plugins
# fix(rbac): reduce the number of times we build group hierarchy graphs

## Hey, I just made a Pull Request!

This PR aims to reduce the number of times that we build group hierarchy graphs during evaluation. Originally, there would be a minimum of three graphs built for each call to `handle` and this would scale up for every role that a single user is attached to, find example below. These changes attempts to reduce this number down to an average of one graph built per eval, see caveat down below for more info.

In these changes, we inside use `getRolesForUser` from casbin to get the direct and indirect roles for the user. We then use these roles and the permission policy that we are evaluating for to filter for just the roles and policies that need to be checked. With the fact that we filter before enforcement, we can assume that `BackstageRoleManager.hasLink` will always be true. There are moments where this isn't always the case and more info can be found in the caveat.

Ex. `user:default/adam` who is a part of three roles. During eval we originally filtered for those roles and casbin would call `BackstageRoleManager.hasLink` to figure out if there was a link or not. This would be done three times, for each role, and the same graph would be built all three times. 

A caveat, the reason on average it should be one graph built is because in the event that a user or group is directly assigned to a permission, `group:default/some-group, catalog-entity, read, allow` where `user:default/some-user` is a part of the group, we end up still resorting to building the graph because of the lack of role assignment. This should ultimately be an outlier since it goes against RBAC but we still need to account for this scenario. So, we will still build the graph in `BackstageRoleManager.hasLink` for eval.

## How to test

This is a big change and I dislike removing test cases since they are no longer valid anymore. However, this doesn't mean that we don't still need those tests. Just that it would be probably better captured in a e2e scenario that will also need to be built and added.

In the meantime, I included the resources to hopefully replicate the tests that were removed in `examples/manual-tests/rbac/`. In this folder is a set or groups, users, roles, and permissions that can be loaded in that should capture all of the tests that were removed. I also included a text file that outlines the different structures of all of the groups. Within the csv that holds the roles and permissions are the different tests that I checked against.

I found that it was easiest to configure a Keycloak instance and set up the OIDC auth provider to test all of the scenarios. 

### Edit

I am leaving the original above, but I was able to include tests for just about every scenario that I could think of. This has started to turn into a large update, but most of it includes the re-added tests (albeit to a new file) and a lot of example catalog entities to make testing manually easier.

#### :heavy_check_mark: Checklist




/cc @AndrienkoAleksandr @dzemanov
